[Hardening] Make Vector::uncheckedAppend() an alias to Vector::append()

cdumez opened the pull request:
#### 1b624bea4bbbf7cf1c9a7b8fc65e9fa6d153d32c
<pre>
[Hardening] Make Vector::uncheckedAppend() an alias to Vector::append()
<a href="https://bugs.webkit.org/show_bug.cgi?id=262431">https://bugs.webkit.org/show_bug.cgi?id=262431</a>

Reviewed by Ryosuke Niwa.

Make Vector::uncheckedAppend() an alias to Vector::append() so that bounds
checking always happens. This hardening is part of our effort to make our code
safer.

This tested as performance neutral on Speedometer, MotionMark and JetStream on
various A/B bots (I tried to cover various configurations). However, I plan to
monitor the bots after landing to be safe. If it sticks, I&apos;ll follow-up to get
rid of Vector::uncheckedAppend() entirely.

Note that in order to avoid regressions on benchmarks, WTF::map(), Vector::map(),
copyToVector() &amp; copyToVectorOf() still rely on a private
unsafeAppendWithoutCapacityCheck() function.

* Source/JavaScriptCore/API/JSObjectRef.cpp:
(JSObjectCopyPropertyNames):
* Source/JavaScriptCore/dfg/DFGLivenessAnalysisPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSSAConversionPhase.cpp:
(JSC::DFG::SSAConversionPhase::run):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::linkOSREntries):
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::sortLocaleData):
(JSC::IntlCollator::searchLocaleData):
* Source/WTF/wtf/Liveness.h:
(WTF::Liveness::compute):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
(WTF::Vector::append):
(WTF::Vector::uncheckedAppend):
(WTF::Vector::uncheckedConstructAndAppend):
(WTF::Vector::unsafeAppendWithoutCapacityCheck):
(WTF::Malloc&gt;::unsafeAppendWithoutCapacityCheck):
(WTF::Malloc&gt;::appendVector):
(WTF::Malloc&gt;::map const const):
(WTF::Mapper::map):
(WTF::copyToVectorSpecialization):
(WTF::Malloc&gt;::uncheckedAppend): Deleted.
(WTF::Malloc&gt;::uncheckedConstructAndAppend): Deleted.
* Source/WTF/wtf/WeakHashSet.h:
* Source/WTF/wtf/WeakListHashSet.h:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::charactersWithoutNullTermination const):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::maybeCreateIDBKeyFromScriptValueAndKeyPath):
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::setUnicodeRange):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didAssociateFormControlsTimerFired):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setElementsArrayAttribute):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::convertNodesOrStringsIntoNode):
* Source/WebCore/html/URLSearchParams.cpp:
(WebCore::URLSearchParams::getAll const):
* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::initializeAttributes):
* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::forEachResource):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::absoluteRectsForRangeInText):
(WebCore::RenderObject::absoluteTextRects):

Canonical link: <a href="https://commits.webkit.org/268700@main">https://commits.webkit.org/268700@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f5fb93af5f1b9bb3ea97c83c66373a66bae5a83

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22306 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19040 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24093 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21008 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20441 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20627 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20499 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17740 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23157 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17667 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18545 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24830 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/17763 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18745 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18721 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22770 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/19809 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19308 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16394 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/23815 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18523 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5702 "Found 4 jsc stress test failures: stress/sampling-profiler-microtasks.js.bytecode-cache, stress/sampling-profiler-microtasks.js.dfg-eager, stress/sampling-profiler-microtasks.js.eager-jettison-no-cjit, stress/sampling-profiler-microtasks.js.mini-mode") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4906 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22860 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/25072 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19132 "Built successfully") | | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5525 "Found 1 jsc stress test failure: microbenchmarks/array-from-object.js.lockdown") | 
<!--EWS-Status-Bubble-End-->

johnwilander reviewed (commented):
I’m not familiar with the intricacies of this code so I shouldn’t formally review it. But I sure am supportive of this change!

cdumez committed:
[Hardening] Make Vector::uncheckedAppend() an alias to Vector::append()
https://bugs.webkit.org/show_bug.cgi?id=262431

Reviewed by Ryosuke Niwa.

Make Vector::uncheckedAppend() an alias to Vector::append() so that bounds
checking always happens. This hardening is part of our effort to make our code
safer.

This tested as performance neutral on Speedometer, MotionMark and JetStream on
various A/B bots (I tried to cover various configurations). However, I plan to
monitor the bots after landing to be safe. If it sticks, I'll follow-up to get
rid of Vector::uncheckedAppend() entirely.

Note that in order to avoid regressions on benchmarks, WTF::map(), Vector::map(),
copyToVector() & copyToVectorOf() still rely on a private
unsafeAppendWithoutCapacityCheck() function.

* Source/JavaScriptCore/API/JSObjectRef.cpp:
(JSObjectCopyPropertyNames):
* Source/JavaScriptCore/dfg/DFGLivenessAnalysisPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSSAConversionPhase.cpp:
(JSC::DFG::SSAConversionPhase::run):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::linkOSREntries):
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::sortLocaleData):
(JSC::IntlCollator::searchLocaleData):
* Source/WTF/wtf/Liveness.h:
(WTF::Liveness::compute):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
(WTF::Vector::append):
(WTF::Vector::uncheckedAppend):
(WTF::Vector::uncheckedConstructAndAppend):
(WTF::Vector::unsafeAppendWithoutCapacityCheck):
(WTF::Malloc>::unsafeAppendWithoutCapacityCheck):
(WTF::Malloc>::appendVector):
(WTF::Malloc>::map const const):
(WTF::Mapper::map):
(WTF::copyToVectorSpecialization):
(WTF::Malloc>::uncheckedAppend): Deleted.
(WTF::Malloc>::uncheckedConstructAndAppend): Deleted.
* Source/WTF/wtf/WeakHashSet.h:
* Source/WTF/wtf/WeakListHashSet.h:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::charactersWithoutNullTermination const):
* Source/WebCore/bindings/js/IDBBindingUtilities.cpp:
(WebCore::maybeCreateIDBKeyFromScriptValueAndKeyPath):
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::setUnicodeRange):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didAssociateFormControlsTimerFired):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setElementsArrayAttribute):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::convertNodesOrStringsIntoNode):
* Source/WebCore/html/URLSearchParams.cpp:
(WebCore::URLSearchParams::getAll const):
* Source/WebCore/html/parser/AtomHTMLToken.h:
(WebCore::AtomHTMLToken::initializeAttributes):
* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::forEachResource):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::absoluteRectsForRangeInText):
(WebCore::RenderObject::absoluteTextRects):

Canonical link: https://commits.webkit.org/268700@main
